### PR TITLE
Feature/orca 465

### PR DESCRIPTION
**Pull request by rizbihassan**
## Summary of Changes

Addresses [ORCA-465: Create or modify deployment script to automate ORCA buckets and dynamodb table creation for running the bamboo.](https://bugs.earthdata.nasa.gov/browse/ORCA-465)

## Changes

* updated bin/deployment-rds.sh to include s3 buckets and dynamo deployment

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

ran the script in bamboo when 
- there were no resources present https://ci.earthdata.nasa.gov/browse/ORCA-PN-DRC-52/log
- when the resources were already present. https://ci.earthdata.nasa.gov/browse/ORCA-PN-DRC-56/log
- when some resources were modified https://ci.earthdata.nasa.gov/browse/ORCA-PN-DRC-55/log

## Checklist:

- [ x] My code follows the style guidelines of this project
- [ x] I have performed a self-review of my own code
- [ x] I have commented my code, particularly in hard-to-understand areas
- [ x] My changes generate no new warnings

### CODE REVIEW

**Review by AndrewDorn — APPROVED** [review]
I like including this in the TF to automatically tie into the `terraform destroy`, especially since that's how we want to eventually create these buckets anyway.

I do think we should put a card on the backlog to have the Orca buckets created in a different account. I understand that's outside the scope of initial implementation.

**Comment by rizbihassan** [comment]
> I like including this in the TF to automatically tie into the `terraform destroy`, especially since that's how we want to eventually create these buckets anyway.
> 
> I do think we should put a card on the backlog to have the Orca buckets created in a different account. I understand that's outside the scope of initial implementation.

We can actually create the buckets in any account from the bamboo pipeline.

When you run the pipeline, you add the aws_access_key and secret_access_key which tells bamboo which aws account to deploy the buckets to.

**Comment by AndrewDorn** [comment]
> > I like including this in the TF to automatically tie into the `terraform destroy`, especially since that's how we want to eventually create these buckets anyway.
> > I do think we should put a card on the backlog to have the Orca buckets created in a different account. I understand that's outside the scope of initial implementation.
> 
> We can actually create the buckets in any account from the bamboo pipeline.
> 
> When you run the pipeline, you add the aws_access_key and secret_access_key which tells bamboo which aws account to deploy the buckets to.

Doesn't that move everything though? For a realistic test environment, the Orca buckets should be in a separate account from stuff like our lambdas.

**Comment by rizbihassan** [comment]
> > > I like including this in the TF to automatically tie into the `terraform destroy`, especially since that's how we want to eventually create these buckets anyway.
> > > I do think we should put a card on the backlog to have the Orca buckets created in a different account. I understand that's outside the scope of initial implementation.
> > 
> > 
> > We can actually create the buckets in any account from the bamboo pipeline.
> > When you run the pipeline, you add the aws_access_key and secret_access_key which tells bamboo which aws account to deploy the buckets to.
> 
> Doesn't that move everything though? For a realistic test environment, the Orca buckets should be in a separate account from stuff like our lambdas.

hmm true it moves everything to a new account.

Do you know if we want to deploy all the buckets to a separate account or only prefix-orca-primary?

**Comment by AndrewDorn** [comment]
> > > > I like including this in the TF to automatically tie into the `terraform destroy`, especially since that's how we want to eventually create these buckets anyway.
> > > > I do think we should put a card on the backlog to have the Orca buckets created in a different account. I understand that's outside the scope of initial implementation.
> > > 
> > > 
> > > We can actually create the buckets in any account from the bamboo pipeline.
> > > When you run the pipeline, you add the aws_access_key and secret_access_key which tells bamboo which aws account to deploy the buckets to.
> > 
> > 
> > Doesn't that move everything though? For a realistic test environment, the Orca buckets should be in a separate account from stuff like our lambdas.
> 
> hmm true it moves everything to a new account.
> 
> Do you know if we want to deploy all the buckets to a separate account or only prefix-orca-primary?

Also the reports bucket. Then they'll need proper permissions for cross-account access. I'm guessing we're getting away with no permissions because they're in the same account.

**Comment by rizbihassan** [comment]
> > > > > I like including this in the TF to automatically tie into the `terraform destroy`, especially since that's how we want to eventually create these buckets anyway.
> > > > > I do think we should put a card on the backlog to have the Orca buckets created in a different account. I understand that's outside the scope of initial implementation.
> > > > 
> > > > 
> > > > We can actually create the buckets in any account from the bamboo pipeline.
> > > > When you run the pipeline, you add the aws_access_key and secret_access_key which tells bamboo which aws account to deploy the buckets to.
> > > 
> > > 
> > > Doesn't that move everything though? For a realistic test environment, the Orca buckets should be in a separate account from stuff like our lambdas.
> > 
> > 
> > hmm true it moves everything to a new account.
> > Do you know if we want to deploy all the buckets to a separate account or only prefix-orca-primary?
> 
> Also the reports bucket. Then they'll need proper permissions for cross-account access. I'm guessing we're getting away with no permissions because they're in the same account.

added orca_reports bucket. This was added in the unreleased version that's why forgot to add this.

If we are only planning to deploy prefix-orca-primary to DR account and other buckets in sandbox, then we have to create a new stage in bamboo to deploy that bucket only using that account's aws access keys. Then when we run the deployment pipeline in sandbox account, we will use the aws access keys for that sandbox account. That's what I can think of right now otherwise we have to manually create the bucket and add cross account policy in DR account for now using cli.

**Comment by rizbihassan** [comment]
> > > > > > > I like including this in the TF to automatically tie into the `terraform destroy`, especially since that's how we want to eventually create these buckets anyway.
> > > > > > > I do think we should put a card on the backlog to have the Orca buckets created in a different account. I understand that's outside the scope of initial implementation.
> > > > > > 
> > > > > > 
> > > > > > We can actually create the buckets in any account from the bamboo pipeline.
> > > > > > When you run the pipeline, you add the aws_access_key and secret_access_key which tells bamboo which aws account to deploy the buckets to.
> > > > > 
> > > > > 
> > > > > Doesn't that move everything though? For a realistic test environment, the Orca buckets should be in a separate account from stuff like our lambdas.
> > > > 
> > > > 
> > > > hmm true it moves everything to a new account.
> > > > Do you know if we want to deploy all the buckets to a separate account or only prefix-orca-primary?
> > > 
> > > 
> > > Also the reports bucket. Then they'll need proper permissions for cross-account access. I'm guessing we're getting away with no permissions because they're in the same account.
> > 
> > 
> > added orca_reports bucket. This was added in the unreleased version that's why forgot to add this.
> > If we are only planning to deploy prefix-orca-primary to DR account and other buckets in sandbox, then we have to create a new stage in bamboo to deploy that bucket only using that account's aws access keys. Then when we run the deployment pipeline in sandbox account, we will use the aws access keys for that sandbox account. That's what I can think of right now otherwise we have to manually create the bucket and add cross account policy in DR account for now using cli.
> 
> We'd also have prefix-orca-reports in the DR account, but otherwise sounds right. That's why I think it should be a backlog card.

yup just making sure. 

created https://bugs.earthdata.nasa.gov/browse/ORCA-476

**Review by bhazuka — CHANGES_REQUESTED** [review]
Overall it is looking good but, some items we should talk through as a team.

File: bin/deployment-rds.sh [hunk 1]
```diff
@@ -5,14 +5,77 @@ export AWS_ACCESS_KEY_ID=$bamboo_AWS_ACCESS_KEY_ID
...
+resource "aws_s3_bucket" "internal" {
+  bucket = "PREFIX-internal"
+}
+resource "aws_s3_bucket" "public" {
+  bucket = "PREFIX-public"
+}
+resource "aws_s3_bucket" "private" {
+  bucket = "PREFIX-private"
+}
+resource "aws_s3_bucket" "level0" {
```
[the remark this answers is not in the document]
Comment (rizbihassan) [reply bin/deployment-rds.sh:23]: Do we have an existing glacier bucket where objects will be moved from this level0 bucket?
or do we have to create a new bucket for this?

File: bin/deployment-rds.sh [hunk 2]
```diff
@@ -5,14 +5,77 @@ export AWS_ACCESS_KEY_ID=$bamboo_AWS_ACCESS_KEY_ID
 export AWS_SECRET_ACCESS_KEY=$bamboo_AWS_SECRET_ACCESS_KEY
 export AWS_DEFAULT_REGION=$bamboo_AWS_DEFAULT_REGION
 
-#clone cumulus orca template for deploying cumulus and orca
+#deploy the S3 buckets and dynamoDB table first
+cat << EOF > resources.tf.template
+# create buckets
```
[the remark this answers is not in the document]
Comment (rizbihassan) [reply bin/deployment-rds.sh:10]: added tags to the buckets.

I was initially thinking to use aws cli in the delete script to delete objects and then delete the buckets while doing the destroy but I guess adding `force_destroy = true` makes sense here. Added.

File: bin/deployment-rds.sh [hunk 3]
```diff
@@ -5,14 +5,77 @@ export AWS_ACCESS_KEY_ID=$bamboo_AWS_ACCESS_KEY_ID
...
+resource "aws_s3_bucket" "level0" {
+  bucket = "PREFIX-level0"
+}
+resource "aws_s3_bucket" "orca-primary" {
+  bucket = "PREFIX-orca-primary"
+}
+resource "aws_s3_bucket" "protected" {
+  bucket = "PREFIX-protected"
+}
+resource "aws_s3_bucket" "orca-reports" {
```
[the remark this answers is not in the document]
Comment (rizbihassan) [reply bin/deployment-rds.sh:32]: yes we can add a rule to delete after 30 days. Will note it in ORCA-476

File: bin/deployment-rds.sh [hunk 4]
```diff
@@ -5,14 +5,77 @@ export AWS_ACCESS_KEY_ID=$bamboo_AWS_ACCESS_KEY_ID
 export AWS_SECRET_ACCESS_KEY=$bamboo_AWS_SECRET_ACCESS_KEY
 export AWS_DEFAULT_REGION=$bamboo_AWS_DEFAULT_REGION
 
-#clone cumulus orca template for deploying cumulus and orca
+#deploy the S3 buckets and dynamoDB table first
+cat << EOF > resources.tf.template
```
[the remark this answers is not in the document]
Comment (rizbihassan) [reply bin/deployment-rds.sh:9]: yes that could be another option. Put this under something like `deployment_templates/` in our repo

File: website/docs/developer/development-guide/code/versioning-releases.md [hunk 5]
```diff
@@ -121,45 +121,13 @@ While running the `Deploy Dev RDS Stack` stage, replace the following variables
 - DB_ADMIN_PASSWORD
 - DB_USER_PASSWORD
 
-Make sure the proper buckets, dynamoDB table and the EC2 key pair are first created manually using the same prefix before running the pipeline. See this cumulus [documentation](https://nasa.github.io/cumulus/docs/deployment/deployment-readme) and [deployment with cumulus documentation](https://github.com/nasa/cumulus-orca/blob/develop/website/docs/developer/deployment-guide/deployment-with-cumulus.md#modifying-cumulus-tfterraformtfvars) for additional details. Typically, these are the buckets that need to be created first
```
[the remark this answers is not in the document]
Comment (rizbihassan) [reply website/docs/developer/development-guide/code/versioning-releases.md:124]: yes it won't hurt to keep them. Added back.

File: bin/deployment-rds.sh [hunk 6]
```diff
@@ -5,14 +5,77 @@ export AWS_ACCESS_KEY_ID=$bamboo_AWS_ACCESS_KEY_ID
...
+    echo "terraform state file not present in S3. Creating the buckets and dynamoDB table..."
+else
+    echo "Terraform state file found in S3 bucket. Using S3 remote backend"
+fi
+
+terraform init -input=false
+echo "Deploying S3  buckets and dynamoDB table"
+terraform apply \
+  -auto-approve \
+  -lock=false \
```
[the remark this answers is not in the document]
Comment (rizbihassan) [reply bin/deployment-rds.sh]: oops removed. Added that while testing

File: bin/deployment-rds.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (rizbihassan) [reply bin/deployment-rds.sh:43]: add a lifecycle config similar to glacier bucket in cumulus OU account.

File: bin/deployment-rds.sh [hunk 7]
```diff
@@ -5,14 +5,77 @@ export AWS_ACCESS_KEY_ID=$bamboo_AWS_ACCESS_KEY_ID
...
+resource "aws_s3_bucket" "public" {
+  bucket = "PREFIX-public"
+}
+resource "aws_s3_bucket" "private" {
+  bucket = "PREFIX-private"
+}
+resource "aws_s3_bucket" "level0" {
+  bucket = "PREFIX-level0"
+}
+resource "aws_s3_bucket" "orca-primary" {
```
[the remark this answers is not in the document]
Comment (rizbihassan) [reply bin/deployment-rds.sh]: deploy WORM bucket in DR account

File: integration_test/deployment-rds.sh [hunk 8]
```diff
@@ -5,54 +5,12 @@ export AWS_ACCESS_KEY_ID=$bamboo_AWS_ACCESS_KEY_ID
...
-    name = "LockID"
-    type = "S"
- }
-}
-EOF
-
 #remove old files from bamboo as they throw error
 rm *.tf
+#deploy the S3 buckets and dynamoDB table first
+git clone --branch develop --single-branch https://github.com/nasa/cumulus-orca.git && cd integration_test
```
[the remark this answers is not in the document]
Comment (rizbihassan) [reply integration_test/deployment-rds.sh:11]: bamboo throws an error without cloning the repo.
the `rds-deployment.sh` did not clone the repo before but we need this now to use the integration_test/buckets.tf.template

**Review by AndrewDorn — APPROVED** [review]

**Review by bhazuka — APPROVED** [review]

**Pull request merged**